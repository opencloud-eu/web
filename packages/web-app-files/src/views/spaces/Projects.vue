<template>
  <div class="oc-flex oc-width-1-1">
    <files-view-wrapper>
      <app-bar
        :breadcrumbs="breadcrumbs"
        :show-actions-on-selection="true"
        :has-bulk-actions="true"
        :has-hidden-files="false"
        :has-file-extensions="false"
        :has-pagination="false"
        :batch-actions-loading="batchActionsLoading"
        :is-side-bar-open="isSideBarOpen"
        :view-modes="viewModes"
        :view-mode-default="FolderViewModeConstants.name.tiles"
      >
        <template #actions>
          <create-space v-if="hasCreatePermission" class="oc-mr-s" />
          <div v-if="!selectedResourcesIds?.length" class="oc-flex oc-flex-middle oc-pl-s">
            <span v-text="$gettext('Learn about spaces')" />
            <oc-contextual-helper
              :list="spacesHelpList"
              :title="$gettext('Spaces')"
              class="oc-ml-xs"
            />
          </div>
        </template>
      </app-bar>
      <app-loading-spinner v-if="areResourcesLoading" />
      <template v-else>
        <no-content-message
          v-if="!spaces.length"
          id="files-spaces-empty"
          class="files-empty"
          icon="layout-grid"
        >
          <template #message>
            <span v-text="$gettext('You don\'t have access to any spaces')" />
          </template>
        </no-content-message>
        <div v-else class="spaces-list">
          <div
            class="spaces-list-filters oc-flex oc-flex-right oc-flex-wrap oc-flex-bottom oc-mx-m oc-mb-m"
          >
            <oc-text-input
              id="spaces-filter"
              v-model="filterTerm"
              :label="$gettext('Search')"
              autocomplete="off"
            />
          </div>
          <component
            :is="folderView.component"
            v-model:selected-ids="selectedResourcesIds"
            resource-type="space"
            :resources="paginatedItems"
            :fields-displayed="tableDisplayFields"
            :sort-fields="sortFields"
            :sort-by="sortBy"
            :sort-dir="sortDir"
            :header-position="fileListHeaderY"
            :is-side-bar-open="isSideBarOpen"
            :view-size="viewSize"
            v-bind="folderView.componentAttrs?.()"
            @sort="handleSort"
            @item-visible="loadPreview({ space: $event, resource: $event })"
          >
            <template #image="{ resource }">
              <template v-if="viewMode === FolderViewModeConstants.name.tiles">
                <oc-spinner
                  v-if="imagesLoading.includes(resource.id)"
                  :aria-label="$gettext('Space image is loading')"
                />
                <img
                  v-else-if="resource.thumbnail"
                  class="tile-preview"
                  :src="resource.thumbnail"
                  alt=""
                />
              </template>
              <template v-else>
                <oc-spinner
                  v-if="imagesLoading.includes(resource.id)"
                  :aria-label="$gettext('Space image is loading')"
                  class="oc-mr-s"
                />
                <img
                  v-else-if="resource.thumbnail"
                  class="table-preview oc-mr-s"
                  :class="{ 'table-preview-disabled': resource.disabled }"
                  :src="resource.thumbnail"
                  alt=""
                  width="33"
                  height="33"
                />
                <resource-icon v-else class="oc-mr-s" :resource="resource" />
              </template>
            </template>
            <template #actions="{ resource }">
              <oc-button
                v-if="!resource.disabled"
                v-oc-tooltip="showSpaceMemberLabel"
                class="spaces-list-show-members-button raw-hover-surface oc-p-xs oc-ml-xs"
                :aria-label="showSpaceMemberLabel"
                appearance="raw"
                @click="openSidebarSharePanel(resource as SpaceResource)"
              >
                <oc-icon name="group" fill-type="line" />
              </oc-button>
            </template>
            <template #contextMenu="{ resource, isOpen, dropRef }">
              <space-context-actions
                v-if="isOpen && isResourceInSelection(resource)"
                :loading="resource.graphPermissions === undefined"
                :action-options="{ resources: [resource] as SpaceResource[] }"
                :drop-ref="dropRef"
              />
            </template>
            <template #footer>
              <pagination :pages="totalPages" :current-page="currentPage" />
              <div class="oc-text-center oc-width-1-1 oc-my-s">
                <p class="oc-text-muted">{{ footerTextTotal }}</p>
                <p v-if="filterTerm" class="oc-text-muted">{{ footerTextFilter }}</p>
              </div>
            </template>
            <!--- table -->
            <template #totalQuota="{ resource }">
              {{ getTotalQuota(resource) }}
            </template>
            <template #usedQuota="{ resource }"> {{ getUsedQuota(resource) }}</template>
            <template #remainingQuota="{ resource }"> {{ getRemainingQuota(resource) }}</template>
          </component>
        </div>
      </template>
    </files-view-wrapper>
    <file-side-bar
      :is-open="isSideBarOpen"
      :active-panel="sideBarActivePanel"
      :space="selectedSpace"
    />
  </div>
</template>

<script setup lang="ts">
import { onMounted, computed, unref, ref, watch, nextTick, onBeforeUnmount } from 'vue'
import { useTask } from 'vue-concurrency'
import Mark from 'mark.js'
import Fuse from 'fuse.js'

import {
  AppLoadingSpinner,
  useResourcesStore,
  useSpacesStore,
  useExtensionRegistry,
  queryItemAsString,
  useLoadPreview
} from '@opencloud-eu/web-pkg'

import { AppBar, CreateSpace } from '@opencloud-eu/web-pkg'
import {
  useAbility,
  useClientService,
  FolderViewModeConstants,
  useRouteQueryPersisted,
  useSort,
  useRouteName,
  usePagination,
  useRouter,
  useRoute,
  Pagination,
  FileSideBar,
  NoContentMessage
} from '@opencloud-eu/web-pkg'
import SpaceContextActions from '../../components/Spaces/SpaceContextActions.vue'
import {
  isProjectSpaceResource,
  ProjectSpaceResource,
  SpaceResource
} from '@opencloud-eu/web-client'
import FilesViewWrapper from '../../components/FilesViewWrapper.vue'
import { eventBus } from '@opencloud-eu/web-pkg'
import { SideBarEventTopics, useSideBar } from '@opencloud-eu/web-pkg'
import { sortFields as availableSortFields, translateSortFields } from '@opencloud-eu/web-pkg'
import { defaultFuseOptions, formatFileSize, ResourceIcon } from '@opencloud-eu/web-pkg'
import { useGettext } from 'vue3-gettext'
import { useKeyboardActions } from '@opencloud-eu/web-pkg'
import {
  useKeyboardFileNavigation,
  useKeyboardFileMouseActions,
  useKeyboardFileActions
} from '../../composables/keyboardActions'
import { orderBy } from 'lodash-es'
import { useResourcesViewDefaults } from '../../composables'
import { folderViewsProjectSpacesExtensionPoint } from '../../extensionPoints'
import { storeToRefs } from 'pinia'

const spacesStore = useSpacesStore()
const router = useRouter()
const route = useRoute()
const clientService = useClientService()
const { can } = useAbility()
const language = useGettext()
const { $gettext, $ngettext } = language
const filterTerm = ref('')
const markInstance = ref(undefined)
const resourcesStore = useResourcesStore()
const { imagesLoading } = storeToRefs(spacesStore)
const { isSideBarOpen, sideBarActivePanel } = useSideBar()

const { setSelection, initResourceList, clearResourceList, setAncestorMetaData } = resourcesStore
const { areDisabledSpacesShown } = storeToRefs(resourcesStore)

const loadResourcesTask = useTask(function* (signal) {
  clearResourceList()
  setAncestorMetaData({})
  yield spacesStore.reloadProjectSpaces({
    graphClient: clientService.graphAuthenticated,
    signal
  })
  initResourceList({ currentFolder: null, resources: unref(spaces) })
})

const {
  viewSize,
  fileListHeaderY,
  scrollToResourceFromRoute,
  areResourcesLoading,
  selectedResourcesIds,
  selectedResources,
  isResourceInSelection
} = useResourcesViewDefaults({ loadResourcesTask })

let loadPreviewToken: string = null

const runtimeSpaces = computed(() => {
  return spacesStore.spaces.filter(isProjectSpaceResource) || []
})
const selectedSpace = computed(() => {
  if (
    unref(selectedResources).length === 1 &&
    isProjectSpaceResource(unref(selectedResources)[0])
  ) {
    return unref(selectedResources)[0] as ProjectSpaceResource
  }
  return null
})

const tableDisplayFields = [
  'image',
  'name',
  'totalQuota',
  'usedQuota',
  'remainingQuota',
  'indicators',
  'mdate'
]

const sortFields = translateSortFields(availableSortFields, language)
const {
  sortBy,
  sortDir,
  items: spaces,
  handleSort
} = useSort<SpaceResource>({
  items: runtimeSpaces,
  fields: sortFields
})
const filter = (spaces: Array<ProjectSpaceResource>, filterTerm: string) => {
  if (!unref(areDisabledSpacesShown)) {
    spaces = spaces.filter((space) => space.disabled !== true)
  }

  if (!(filterTerm || '').trim()) {
    return spaces
  }

  const searchEngine = new Fuse(spaces, { ...defaultFuseOptions, keys: ['name'] })
  return searchEngine.search(filterTerm).map((r) => r.item)
}
const items = computed(() => {
  return orderBy(
    filter(unref(spaces), unref(filterTerm)),
    [
      (item: SpaceResource) =>
        typeof item[unref(sortBy)] === 'string'
          ? item[unref(sortBy)].toLowerCase()
          : item[unref(sortBy)]
    ],
    unref(sortDir)
  )
})

const {
  items: paginatedItems,
  page: currentPage,
  total: totalPages
} = usePagination({
  items,
  perPageDefault: '50',
  perPageStoragePrefix: 'spaces-list'
})

const batchActionsLoading = computed(() => {
  const selectedSpaces = unref(selectedResources) as SpaceResource[]
  return selectedSpaces.some(({ graphPermissions }) => graphPermissions === undefined)
})

watch(filterTerm, async () => {
  const instance = unref(markInstance)
  if (!instance) {
    return
  }
  await router.push({ ...unref(route), query: { ...unref(route).query, page: '1' } })
  instance.unmark()
  instance.mark(unref(filterTerm), {
    element: 'span',
    className: 'mark-highlight',
    exclude: ['th *', 'tfoot *']
  })
})

watch(selectedResourcesIds, async (ids) => {
  if (!ids.length) {
    return
  }

  await spacesStore.loadGraphPermissions({
    ids,
    graphClient: clientService.graphAuthenticated
  })
})

const hasCreatePermission = computed(() => can('create-all', 'Drive'))

const extensionRegistry = useExtensionRegistry()
const viewModes = computed(() => {
  return [
    ...extensionRegistry
      .requestExtensions(folderViewsProjectSpacesExtensionPoint)
      .map((e) => e.folderView)
  ]
})

const routeName = useRouteName()

const viewMode = useRouteQueryPersisted({
  name: `${unref(routeName)}-${FolderViewModeConstants.queryName}`,
  defaultValue: FolderViewModeConstants.name.tiles
})
const viewModeStr = computed(() => queryItemAsString(unref(viewMode)))

const { loadPreview } = useLoadPreview(viewModeStr)

const keyActions = useKeyboardActions()
useKeyboardFileNavigation(keyActions, runtimeSpaces, viewMode)
useKeyboardFileMouseActions(keyActions, viewMode)
useKeyboardFileActions(keyActions)

const getTotalQuota = (space: SpaceResource) => {
  if (space.spaceQuota.total === 0) {
    return $gettext('Unrestricted')
  }

  return formatFileSize(space.spaceQuota.total, language.current)
}
const getUsedQuota = (space: SpaceResource) => {
  if (space.spaceQuota.used === undefined) {
    return '-'
  }
  return formatFileSize(space.spaceQuota.used, language.current)
}
const getRemainingQuota = (space: SpaceResource) => {
  if (space.spaceQuota.total === 0) {
    return $gettext('Unrestricted')
  }
  if (space.spaceQuota.remaining === undefined) {
    return '-'
  }
  return formatFileSize(space.spaceQuota.remaining, language.current)
}

onMounted(async () => {
  await loadResourcesTask.perform()

  loadPreviewToken = eventBus.subscribe(
    'app.files.spaces.uploaded-image',
    (space: SpaceResource) => {
      loadPreview({ space, resource: space })
    }
  )
  scrollToResourceFromRoute(unref(spaces), 'files-app-bar')
  nextTick(() => {
    markInstance.value = new Mark('.spaces-table')
  })
})

onBeforeUnmount(() => {
  eventBus.unsubscribe('app.files.spaces.uploaded-image', loadPreviewToken)
})

const footerTextTotal = computed(() => {
  const disabledSpaces = unref(spaces).filter((space) => space.disabled === true)

  if (!disabledSpaces.length) {
    return $ngettext(
      '%{spaceCount} space in total',
      '%{spaceCount} spaces in total',
      unref(spaces).length,
      {
        spaceCount: unref(spaces).length.toString()
      }
    )
  }

  return $ngettext(
    '%{spaceCount} space in total (including %{disabledSpaceCount} disabled)',
    '%{spaceCount} spaces in total (including %{disabledSpaceCount} disabled)',
    unref(spaces).length,
    {
      spaceCount: unref(spaces).length.toString(),
      disabledSpaceCount: disabledSpaces.length.toString()
    }
  )
})
const footerTextFilter = computed(() => {
  return $gettext('%{spaceCount} matching spaces', {
    spaceCount: unref(items).length.toString()
  })
})

const folderView = computed(() => {
  const viewModeName = unref(viewMode) || FolderViewModeConstants.name.tiles
  return unref(viewModes).find((v) => v.name === viewModeName)
})

const spacesHelpList = computed(() => {
  return [
    {
      text: $gettext('Spaces are special folders for making files accessible to a team.')
    },
    {
      text: $gettext(
        'Spaces belong to a team and not to a single person. Even if members are removed, the files remain in the Space so that the team can continue to work on the files.'
      )
    },
    {
      text: $gettext(
        'Members with the Manager role can add or remove other members from the Space.'
      )
    },
    {
      text: $gettext('A Space can have multiple Managers. Each Space has at least one Manager.')
    }
  ]
})

const breadcrumbs = computed(() => {
  return [
    {
      text: $gettext('Spaces'),
      onClick: () => loadResourcesTask.perform(),
      isStativNav: true
    }
  ]
})
const showSpaceMemberLabel = computed(() => {
  return $gettext('Show members')
})

const openSidebarSharePanel = (space: SpaceResource) => {
  setSelection([space.id])
  eventBus.publish(SideBarEventTopics.openWithPanel, 'space-share')
}
</script>

<style lang="scss">
#files-spaces-empty {
  height: 75vh;
}

.table-preview {
  object-fit: cover;
  border-radius: 3px;
}

.table-preview-disabled {
  filter: grayscale(100%);
  opacity: 80%;
}

.state-trashed {
  .tile-preview,
  .tile-default-image > svg {
    filter: grayscale(100%);
    opacity: 80%;
  }
}
</style>
