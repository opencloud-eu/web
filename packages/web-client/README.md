# web-client

The `web-client` is a standalone package that allows you to interact with the [OpenCloud](https://github.com/opencloud-eu/opencloud/) APIs via TypeScript. It provides an abstraction layer between the server and a (web-) application that converts API data into objects with helpful types and utilities. This abstraction ensures that users of the APIs don't need in-depth knowledge about them, such as required methods or returned status codes.

The supported APIs are:

- Graph (drives, sharing, user & group management)
- OCS (capabilities & url signing)
- WebDAV (file operations)

## Installation

Depending on your package manager, run one of the following commands:

```
$ npm install @opencloud-eu/web-client

$ pnpm add @opencloud-eu/web-client

$ yarn add @opencloud-eu/web-client
```

## Usage

### Graph

The graph client needs to be instantiated with a base URI corresponding to your OpenCloud deployment and an axios instance. The axios instance is being used for all requests, which means it needs to include all relevant headers either statically or via interceptor.

```
import axios from axios
import { graph } from '@opencloud-eu/web-client'

const accessToken = 'some_access_token'
const baseURI = 'some_base_uri'

const axiosClient = axios.create({
	headers: { Authorization: accessToken }
})

const graphClient = graph(baseURI, axiosClient)
```

The following example demonstrates how to retrieve all spaces accessible to the user. A `SpaceResource` can then be used to e.g. fetch files and folders (see webdav example down below).

```
const mySpaces = await graphClient.drives.listMyDrives()
```

### OCS

The ocs client needs to be instantiated with a base URI corresponding to your OpenCloud deployment and an axios instance. The axios instance is being used for all requests, which means it needs to include all relevant headers either statically or via interceptor.

```
import axios from axios
import { ocs } from '@opencloud-eu/web-client'

const accessToken = 'some_access_token'
const baseURI = 'some_base_uri'

const axiosClient = axios.create({
	headers: { Authorization: accessToken }
})

const ocsClient = ocs(baseURI, axiosClient)
```

The following examples demonstrate how to fetch capabilities and sign URLs.

```
const capabilities = await ocsClient.getCapabilities()

const signedUrl = await ocsClient.signUrl('some_url_to_sign', 'your_username')
```

### WebDav

The webdav client needs to be instantiated with a base URI corresponding to your OpenCloud deployment. You can also pass a header callback which will be called with every dav request.

```
import { webdav } from '@opencloud-eu/web-client'

const accessToken = 'some_access_token'
const baseURI = 'some_base_uri'

const webDavClient = webdav(
	baseURI,
	() => ({ Authorization: accessToken })
)
```

The following example demonstrates how to list all resources of a given `SpaceResource` (see above how to fetch space resources).

```
const { resource, children } = await webDavClient.listFiles(spaceResource)
```
