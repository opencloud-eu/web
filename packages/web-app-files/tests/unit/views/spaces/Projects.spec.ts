import Projects from '../../../../src/views/spaces/Projects.vue'
import { mock } from 'vitest-mock-extended'
import { h, nextTick, ref } from 'vue'
import {
  queryItemAsString,
  useFileActionsDelete,
  useExtensionRegistry,
  FolderViewExtension
} from '@opencloud-eu/web-pkg'

import {
  defaultPlugins,
  mount,
  defaultComponentMocks,
  defaultStubs,
  RouteLocation,
  PiniaMockOptions
} from '@opencloud-eu/web-test-helpers'
import { AbilityRule, SpaceResource } from '@opencloud-eu/web-client'
import {
  folderViewsFavoritesExtensionPoint,
  folderViewsFolderExtensionPoint,
  folderViewsProjectSpacesExtensionPoint
} from '../../../../src/extensionPoints'

vi.mock('@opencloud-eu/web-pkg', async (importOriginal) => ({
  ...(await importOriginal<any>()),
  displayPositionedDropdown: vi.fn(),
  queryItemAsString: vi.fn(),
  appDefaults: vi.fn(),
  useRouteQueryPersisted: vi.fn().mockImplementation(() => ref('resource-table')),
  useFileActions: vi.fn(),
  useFileActionsDelete: vi.fn(() => mock<ReturnType<typeof useFileActionsDelete>>())
}))

const spacesResources = [
  {
    id: '1',
    name: 'Some space',
    driveType: 'project',
    description: 'desc',
    path: '',
    type: 'space',
    isFolder: true,
    disabled: false,
    getDriveAliasAndItem: () => '1'
  },
  {
    id: '2',
    name: 'Some other space',
    driveType: 'project',
    description: 'desc',
    path: '',
    type: 'space',
    isFolder: true,
    disabled: false,
    getDriveAliasAndItem: () => '2'
  },
  {
    id: '3',
    name: 'Some disabled space',
    driveType: 'project',
    description: 'desc',
    path: '',
    type: 'space',
    isFolder: true,
    disabled: true,
    getDriveAliasAndItem: () => '2'
  }
] as unknown as SpaceResource[]

describe('Projects view', () => {
  it('appBar always present', () => {
    const { wrapper } = getMountedWrapper()
    expect(wrapper.find('app-bar-stub').exists()).toBeTruthy()
  })
  it('sideBar always present', () => {
    const { wrapper } = getMountedWrapper()
    expect(wrapper.find('file-side-bar-stub').exists()).toBeTruthy()
  })
  describe('different files view states', () => {
    it('shows the loading spinner during loading', () => {
      const { wrapper } = getMountedWrapper()
      expect(wrapper.find('oc-spinner-stub').exists()).toBeTruthy()
    })
    it('shows the no-content-message after loading', async () => {
      const { wrapper } = getMountedWrapper()
      await wrapper.vm.loadResourcesTask.last
      expect(wrapper.find('oc-spinner-stub').exists()).toBeFalsy()
      expect(wrapper.find('.no-content-message').exists()).toBeTruthy()
    })
    it('lists all available project spaces', async () => {
      const spaces = spacesResources
      const { wrapper } = getMountedWrapper({ spaces })
      await wrapper.vm.loadResourcesTask.last
      expect(wrapper.html()).toMatchSnapshot()
      expect(wrapper.find('.no-content-message').exists()).toBeFalsy()
      expect(wrapper.find('.spaces-list').exists()).toBeTruthy()
    })
    it('shows only filtered spaces if filter applied', async () => {
      const { wrapper } = getMountedWrapper({ spaces: spacesResources })
      wrapper.vm.filterTerm = 'Some other space'
      await nextTick()
      expect(wrapper.vm.items).toEqual([spacesResources[1]])
    })
    it('shows only enabled spaces if includeDisabled filter is not applied', async () => {
      const { wrapper } = getMountedWrapper({ spaces: spacesResources })
      await nextTick()
      expect(wrapper.vm.items.length).toEqual(2)
    })
    it('shows all spaces if includeDisabled filter is applied', async () => {
      const { wrapper } = getMountedWrapper({ spaces: spacesResources, includeDisabled: true })
      await nextTick()
      expect(wrapper.vm.items.length).toEqual(3)
    })
  })
  it('should display the "Create Space"-button when permission given', () => {
    const { wrapper } = getMountedWrapper({
      abilities: [{ action: 'create-all', subject: 'Drive' }],
      stubAppBar: false
    })
    expect(wrapper.find('create-space-stub').exists()).toBeTruthy()
  })
  it('should not pass selected resource as space to sidebar when driveType is not "project"', () => {
    const resource = mock<SpaceResource>({ id: 'selected-resource', driveType: 'personal' })
    const { wrapper } = getMountedWrapper({
      store: { resourcesStore: { resources: [resource], selectedIds: ['selected-resource'] } }
    })

    expect(wrapper.vm.selectedSpace).toStrictEqual(null)
  })
  it('should pass selected resource as space to sidebar when driveType is "project"', () => {
    const resource = mock<SpaceResource>({ id: 'selected-resource', driveType: 'project' })
    const { wrapper } = getMountedWrapper({
      store: { resourcesStore: { resources: [resource], selectedIds: ['selected-resource'] } }
    })

    expect(wrapper.vm.selectedSpace.id).toStrictEqual('selected-resource')
  })
})

function getMountedWrapper({
  mocks = {},
  spaces = [],
  abilities = [],
  stubAppBar = true,
  includeDisabled = false,
  store = {}
}: {
  mocks?: Record<string, unknown>
  spaces?: SpaceResource[]
  abilities?: AbilityRule[]
  stubAppBar?: boolean
  includeDisabled?: boolean
  store?: PiniaMockOptions
} = {}) {
  const plugins = defaultPlugins({
    abilities,
    piniaOptions: {
      spacesState: { spaces },
      resourcesStore: { areDisabledSpacesShown: includeDisabled },
      ...store
    }
  })

  vi.mocked(queryItemAsString).mockImplementation(() => includeDisabled.toString())

  const extensions = [
    {
      id: 'com.github.opencloud-eu.web.files.folder-view.resource-table',
      type: 'folderView',
      extensionPointIds: [
        folderViewsFolderExtensionPoint.id,
        folderViewsProjectSpacesExtensionPoint.id,
        folderViewsFavoritesExtensionPoint.id
      ],
      folderView: {
        name: 'resource-table',
        label: 'Switch to default view',
        icon: {
          name: 'menu-line',
          fillType: 'none'
        },
        component: h('div', { class: 'resource-table' })
      }
    }
  ] satisfies FolderViewExtension[]
  const { requestExtensions } = useExtensionRegistry()
  vi.mocked(requestExtensions).mockReturnValue(extensions)

  const defaultMocks = {
    ...defaultComponentMocks({
      currentRoute: mock<RouteLocation>({ name: 'files-spaces-projects' })
    }),
    ...(mocks && mocks)
  }

  return {
    mocks: defaultMocks,
    wrapper: mount(Projects, {
      global: {
        plugins,
        mocks: defaultMocks,
        provide: defaultMocks,
        stubs: {
          ...defaultStubs,
          'space-context-actions': true,
          'app-bar': stubAppBar,
          CreateSpace: true
        }
      }
    })
  }
}
