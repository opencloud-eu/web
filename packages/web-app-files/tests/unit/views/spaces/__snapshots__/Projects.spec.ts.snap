// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Projects view > different files view states > lists all available project spaces 1`] = `
"<div class="oc-flex oc-width-1-1">
  <div data-v-39c970be="" class="files-view-wrapper oc-width-expand">
    <div data-v-39c970be="" id="files-view">
      <app-bar-stub viewmodedefault="resource-tiles" breadcrumbs="[object Object]" breadcrumbscontextactionsitems="" viewmodes="[object Object]" hasbulkactions="true" hasviewoptions="true" hashiddenfiles="false" hasfileextensions="false" haspagination="false" showactionsonselection="true" issidebaropen="false"></app-bar-stub>
      <div class="spaces-list">
        <div class="spaces-list-filters oc-flex oc-flex-right oc-flex-wrap oc-flex-bottom oc-mx-m oc-mb-m">
          <div class=""><label class="oc-label" for="spaces-filter">Search</label>
            <div class="oc-position-relative">
              <!--v-if--> <input id="spaces-filter" autocomplete="off" aria-invalid="false" class="oc-text-input oc-input oc-rounded" type="text" value="">
              <!--v-if-->
            </div>
            <!--v-if-->
            <!--v-if-->
          </div>
        </div>
        <table class="oc-table oc-table-hover oc-table-sticky has-item-context-menu spaces-table" sort-fields="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]" view-size="1" id="files-space-table" selection="">
          <thead class="oc-thead">
            <tr class="oc-table-header-row">
              <th class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-shrink oc-th oc-table-header-cell oc-table-header-cell-select oc-pl-m" style="top: 0px;">
                <div><span class="oc-table-thead-content"><div class="resource-table-select-all"><span><input id="resource-table-select-all" type="checkbox" name="checkbox" class="oc-checkbox oc-rounded oc-checkbox-l" aria-label="Select all"> <!--v-if--></span></div></span>
      </div>
      </th>
      <th class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-expand oc-th oc-table-header-cell oc-table-header-cell-name" style="top: 0px;" aria-sort="ascending"><button type="button" aria-label="Sort by name" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-s oc-button-secondary oc-button-raw oc-button-secondary-raw no-hover oc-button-sort oc-width-1-1">
          <!--v-if-->
          <!-- @slot Content of the button --> <span class="oc-table-thead-content header-text">Name</span> <span class="oc-icon oc-icon-s oc-p-xs oc-rounded"><!----></span>
        </button></th>
      <th class="oc-table-cell oc-table-cell-align-right oc-table-cell-align-middle oc-table-cell-width-shrink oc-th oc-table-header-cell oc-table-header-cell-status" style="top: 0px;">
        <div><span class="oc-table-thead-content header-text">Status</span></div>
      </th>
      <th class="oc-table-cell oc-table-cell-align-right oc-table-cell-align-middle oc-table-cell-width-shrink oc-th oc-table-header-cell oc-table-header-cell-actions oc-pr-m" style="top: 0px;">
        <div><span class="oc-table-thead-content header-text">Actions</span></div>
      </th>
      </tr>
      </thead>
      <tbody class="has-item-context-menu">
        <tr class="oc-tbody-tr oc-tbody-tr-2" data-item-id="2" draggable="false">
          <td class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-td" colspan="4"><span class="shimmer"></span></td>
        </tr>
        <tr class="oc-tbody-tr oc-tbody-tr-1" data-item-id="1" draggable="false">
          <td class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-td" colspan="4"><span class="shimmer"></span></td>
        </tr>
      </tbody>
      <tfoot class="oc-table-footer">
        <tr class="oc-table-footer-row">
          <td colspan="4" class="oc-table-footer-cell">
            <!-- @slot Footer of the table -->
            <!-- @slot Footer of the files table -->
            <!--v-if-->
            <div class="oc-text-center oc-width-1-1 oc-my-s">
              <p class="oc-text-muted">3 spaces in total (including 1 disabled)</p>
              <!--v-if-->
            </div>
          </td>
        </tr>
      </tfoot>
      </table>
      <!--v-if-->
    </div>
  </div>
</div>
<!--v-if-->
<file-side-bar-stub isopen="false"></file-side-bar-stub>
</div>"
`;
